internal/keyvalue/memorydb.go

```diff
@@ -0,0 +1,295 @@
+// functionality it also supports batch writes and iterating over the keyspace in
+// binary-alphabetical order.
+type MemoryDatabase struct {
+	db   map[string][]byte
+	lock sync.RWMutex
+}
+
+// NewMemoryDatabase returns a wrapped map with all the required database interface
+// methods implemented.
+func NewMemoryDatabase() *MemoryDatabase {
+	return &MemoryDatabase{
+		db: make(map[string][]byte),
+	}
+}
+
+// NewMemoryDatabaseWithCap returns a wrapped map pre-allocated to the provided
+// capcity with all the required database interface methods implemented.
+func NewMemoryDatabaseWithCap(size int) database.KeyValueStore {
+	return &MemoryDatabase{
+		db: make(map[string][]byte, size),
+	}
+}
+
+// Close deallocates the internal map and ensures any consecutive data access op
+// failes with an error.
+func (db *MemoryDatabase) Close() error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	db.db = nil
+	return nil
+}
+
+// Has retrieves if a key is present in the key-value store.
+func (db *MemoryDatabase) Has(key []byte) (bool, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.db == nil {
+		return false, errMemorydbClosed
+	}
+	_, ok := db.db[string(key)]
+	return ok, nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (db *MemoryDatabase) Get(key []byte) ([]byte, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	if db.db == nil {
+		return nil, errMemorydbClosed
+	}
+	if entry, ok := db.db[string(key)]; ok {
+		return common.CopyBytes(entry), nil
+	}
+	return nil, errMemorydbNotFound
+}
+
+// Put inserts the given value into the key-value store.
+func (db *MemoryDatabase) Put(key []byte, value []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.db == nil {
+		return errMemorydbClosed
+	}
+	db.db[string(key)] = common.CopyBytes(value)
+	return nil
+}
+
+// Delete removes the key from the key-value store.
+func (db *MemoryDatabase) Delete(key []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	if db.db == nil {
+		return errMemorydbClosed
+	}
+	delete(db.db, string(key))
+	return nil
+}
+
+// NewBatch creates a write-only key-value store that buffers changes to its host
+// database until a final write is called.
+func (db *MemoryDatabase) NewBatch() database.Batch {
+	return &memoryBatch{
+		db: db,
+	}
+}
+
+// NewIterator creates a binary-alphabetical iterator over the entire keyspace
+// contained within the memory database.
+func (db *MemoryDatabase) NewIterator() database.Iterator {
+	return db.NewIteratorWithPrefix(nil)
+}
+
+// NewIteratorWithPrefix creates a binary-alphabetical iterator over a subset
+// of database content with a particular key prefix.
+func (db *MemoryDatabase) NewIteratorWithPrefix(prefix []byte) database.Iterator {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	var (
+		pr     = string(prefix)
+		keys   = []string{"FAKE KEY TO SIMPLIFY NEXT"}
+		values = [][]byte{[]byte("FAKE VAL TO SIMPLIFY NEXT")}
```

Comment:
This looks very much like code left in from a debugging session. If that's not the case, perhaps some more docs could be added?

Reply:
Iteration works by initially placing the iterator *before* the first item, requiring a call to `Next` to actually advance to the first one. This special case can either be done explicitly within the iterator code, or in my case I made a hack whereby there's actually a 0-th dummy element that the iterator is positioned on when created.

Reply:
An alternative code would be

```go
diff --git a/internal/keyvalue/memorydb.go b/internal/keyvalue/memorydb.go
index 453eba0a0..acee1dcc5 100644
--- a/internal/keyvalue/memorydb.go
+++ b/internal/keyvalue/memorydb.go
@@ -142,8 +142,8 @@ func (db *MemoryDatabase) NewIteratorWithPrefix(prefix []byte) database.Iterator
 
        var (
                pr     = string(prefix)
-               keys   = []string{"FAKE KEY TO SIMPLIFY NEXT"}
-               values = [][]byte{[]byte("FAKE VAL TO SIMPLIFY NEXT")}
+               keys   = make([]string, 0, len(db.db))
+               values = make([][]byte, 0, len(db.db))
        )
        // Collect the keys from the memory database corresponding to the given prefix
        for key := range db.db {
@@ -152,8 +152,8 @@ func (db *MemoryDatabase) NewIteratorWithPrefix(prefix []byte) database.Iterator
                }
        }
        // Sort the items and retrieve the associated values
-       sort.Strings(keys[1:])
-       for _, key := range keys[1:] {
+       sort.Strings(keys)
+       for _, key := range keys {
                values = append(values, db.db[key])
        }
        return &memoryIterator{
@@ -178,22 +178,6 @@ func (db *MemoryDatabase) Len() int {
        return len(db.db)
 }
 
-/*
-// Keys retrieves a copy of all the keys present in the database.
-//
-// Note, this method is only used for testing (i.e. not public in general) and
-// does not have explicit checks for closed-ness to allow simpler testing code.
-func (db *MemoryDatabase) Keys() [][]byte {
-       db.lock.RLock()
-       defer db.lock.RUnlock()
-
-       keys := [][]byte{}
-       for key := range db.db {
-               keys = append(keys, []byte(key))
-       }
-       return keys
-}*/
-
 // keyvalue is a key-value tuple tagged with a deletion field to allow creating
 // memory-database write batches.
 type keyvalue struct {
@@ -254,6 +238,7 @@ func (b *memoryBatch) Reset() {
 // key value store. Internally it is a deep copy of the entire iterated state,
 // sorted by keys.
 type memoryIterator struct {
+       inited bool
        keys   []string
        values [][]byte
 }
@@ -261,6 +246,12 @@ type memoryIterator struct {
 // Next moves the iterator to the next key/value pair. It returns whether the
 // iterator is exhausted.
 func (it *memoryIterator) Next() bool {
+       // If the iterator was not yet initialized, do it now
+       if !it.inited {
+               it.inited = true
+               return len(it.keys) > 0
+       }
+       // Iterator already initialize, advance it
        if len(it.keys) > 0 {
                it.keys = it.keys[1:]
                it.values = it.values[1:]
```

Perhaps it's better to keep it clean vs. dummy values.

Reply:
I don't really care, just wanted to check that it wasn't just some stale leftover

Reply:
Please see my last commit. I've swapped out the dummy items with special case handling.